node/bindnode/node.go

```diff
@@ -432,6 +442,27 @@ func (w *_node) AsBytes() ([]byte, error) {
 	if err := compatibleKind(w.schemaType, datamodel.Kind_Bytes); err != nil {
 		return nil, err
 	}
+
+	typ := w.val.Type()
+	if w.val.Kind() == reflect.Ptr {
+		typ = typ.Elem()
+	}
+	customConverter, ok := w.cfg.customConverters[typ]
+	if ok {
+		res := customConverter.toFunc.Call([]reflect.Value{w.val.Addr()})
+		if !res[1].IsNil() {
+			err, ok := res[1].Interface().(error)
+			if !ok {
+				return nil, fmt.Errorf("converter function did not return expected error: %v", res[1].Interface())
+			}
+			return nil, err
+		}
+		byts, ok := res[0].Interface().([]byte)
```

Comment:
having to do the call in the abstract reflection world is going to be slower than if you can use a concrete function - i wonder if generics could be enough for defining these functions rather than needing to have it as a `reflect.Value`

Reply:
yeah, I wasn't aware of the size of that cost .. looks like more than 70 times slower than a plain call. Which is not awesome. It's not the worst thing in the world but if we're trying to compete against cbor-gen we're already starting off on a back foot.

So a couple of options:

1. Use `interface{}` where for the custom type, and pass in that type via other means (probably just a null pointer like the rest of bindnode uses). Internally it can cast through the various options to find the one that is being used (e.g. `FromBytes([]byte) (interface{}, error)` & `ToBytes(interface{}) ([]byte, error)`. Then the API just needs that extra argument to pass the type. This steps back a little to the previous version of this API (12a0ca48c); it's not terrible but more verbose is slightly annoying.
2. Generics - this is the nice use-case for generics, but I don't believe we can even use reflection to access the concrete type being used, so we still need to derive that from somewhere. So we end up very similar to option 1, of needing to add the type. I think also we might have trouble casting to find out which one is being used and may have to use reflection to figure that out (like the current impl) but I'm speaking out of a bit of ignorance on what Go generics are giving us.
3. Live with the cost and 🤞 that https://github.com/golang/go/issues/49340 gets implemented sometime soon.

I'll try option 1 out and see how it goes. I'm making this API as "EXPERIMENTAL" in its docs to give a bit of permission to tweak and change it as we figure out how well it works for what we need.